iceoryx_hoofs/test/moduletests/test_containers_uninitialized_array.cpp

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "test.hpp"
+
+#include "iceoryx_hoofs/containers/uninitialized_array.hpp"
+
+namespace
+{
+using namespace ::testing;
+
+using iox::containers::UninitializedArray;
+
+template <typename T>
+class UninitializedArrayTest : public ::testing::Test
+{
+  public:
+    using Buffer = T;
+    Buffer buffer{};
+
+    void fillBuffer(int startValue)
+    {
+        auto capacity = buffer.capacity();
+
+        int value = startValue;
+        for (uint64_t i = 0; i < capacity; ++i)
+        {
+            buffer[i] = value++;
```

Comment:
In the context in which you are using the uninitialized array this is undefined behavior since the element at position `i` in the buffer needs to be constructed first with placement new.
Could you therefore adjust the code to:
```
new (&buffer[i]) Type(value++);
```

To make this easily usable I would add another member type alias called `value_type` to the underlying value directly in the uninitialized array. The STL containers are also providing such type aliases to make them simpler to use in templated code, see: https://en.cppreference.com/w/cpp/container/array section `Member types`

With this you can add the alias `using Type = typename Buffer::value_type` in the `UninitializedArrayTest` which provides you access to the actual type.

Reply:
One could also think about a variadic `emplace(index, args)` methods to initialize the value at index which does call placement new. However, we should not track whether it was already initialized as it is a low level construct and hence it should be the responsibility of the user.

Reply:
Such an `emplace` method would be nice.

Reply:
Well, I have no hard feelings against such an `emplace` method but the current implementation aims to be a very low-level building block. If we want to have `emplace` how should it behave when for instance `index` is out of bounds? A check doesn't feel "low-level" anymore, and the number of undefined behavior methods I want to keep to a minimum. I'm a bit torn here but tend to refrain from providing this method.